src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -546,40 +546,34 @@ void ShenandoahOldHeuristics::clear_triggers() {
+                                                               size_t old_region_count, size_t num_regions) {
   if (ShenandoahGenerationalHumongousReserve > 0) {
     size_t old_region_span = (first_old_region <= last_old_region)? (last_old_region + 1 - first_old_region): 0;
     size_t allowed_old_gen_span = num_regions - (ShenandoahGenerationalHumongousReserve * num_regions) / 100;
 
-    // Tolerate lower density if total span is small.  Here's the implementation:
-    //   if old_gen spans more than 100% and density < 75%, trigger old-defrag
-    //   else if old_gen spans more than 87.5% and density < 62.5%, trigger old-defrag
-    //   else if old_gen spans more than 75% and density < 50%, trigger old-defrag
-    //   else if old_gen spans more than 62.5% and density < 37.5%, trigger old-defrag
-    //   else if old_gen spans more than 50% and density < 25%, trigger old-defrag
-    //
-    // A previous implementation was more aggressive in triggering, resulting in degraded throughput when
-    // humongous allocation was not required.
-
-    size_t old_available = _old_gen->available();
-    size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
-    size_t old_unaffiliated_available = _old_gen->free_unaffiliated_regions() * region_size_bytes;
+    size_t old_available = _old_gen->available() / HeapWordSize;
+    size_t region_size_words = ShenandoahHeapRegion::region_size_words();
+    size_t old_unaffiliated_available = _old_gen->free_unaffiliated_regions() * region_size_words;
     assert(old_available >= old_unaffiliated_available, "sanity");
     size_t old_fragmented_available = old_available - old_unaffiliated_available;
 
-    size_t old_bytes_consumed = old_region_count * region_size_bytes - old_fragmented_available;
-    size_t old_bytes_spanned = old_region_span * region_size_bytes;
-    double old_density = ((double) old_bytes_consumed) / old_bytes_spanned;
-
-    uint eighths = 8;
-    for (uint i = 0; i < 5; i++) {
-      size_t span_threshold = eighths * allowed_old_gen_span / 8;
-      double density_threshold = (eighths - 2) / 8.0;
-      if ((old_region_span >= span_threshold) && (old_density < density_threshold)) {
-        trigger_old_is_fragmented(old_density, first_old_region, last_old_region);
-        return;
-      }
-      eighths--;
+    size_t old_words_consumed = old_region_count * region_size_words - old_fragmented_available;
+    size_t old_words_spanned = old_region_span * region_size_words;
+    double old_density = ((double) old_words_consumed) / old_words_spanned;
+
+    double old_span_percent = ((double) old_region_span) / allowed_old_gen_span;
+    double old_span_percent_squared = old_span_percent * old_span_percent;
+
+    if ((old_span_percent >= 0.50) && (old_density / old_span_percent_squared < 0.75)) {
+      // We trigger old defragmentation, for example, if:
+      //  old_span_percent is 100% and old_density is below 75.0%, or
```

Comment:
Note that old_span_percent may be greater than 100%.  Can add comment:
  old_span_percent is 110% and old_density is below 90.8%, or
  
At higher existing utilization, there is less benefit to defragmenting, because there are diminishing returns for the larger efforts.  We're just copying nearly full regions around...

Reply:
But if we are at 110% old_span_percent, we have violated our intended Humongous Reserves, so I'm thinking we should try to squeeze that last 10% out of old so we don't have to STW on the next humongous allocation request...

Reply:
I have a few broader questions:
* why is `allowed_old_gen_span` defined in terms of humongous reserve?
* why do we want to square `old_span_percent` and divide it into `old_density`? The comments suggest we ought to be comparing `old_density` directly?

Reply:
The idea is that services that anticipate a need for "lots of humongous" allocation can specify a non-zero value for ShenandoahGenerationalHumongousReserve.  If this value is non-zero, GenShen endeavors to avoid placing old-gen regions into the first N% of the heap.  The low end of the heap is reserved for humongous regions, leaving the rest of the heap eligible for Old regions.

Each time we rebuild the freeset, the top-most available regions are reserved for old collector. Next to these, we reserve for the collector (young "survivor regions").  The remaining heap is available for mutator allocations, with humongous allocations prioritized to bottom.

There's an "implicit assumption" in this code that the top-most old-gen region is aligned at top of the heap.  I'll change this by defining old_region_span as (num_regions() - first_old_region).  This makes the correlation more clear.  (I'll also add some comments).

The reason for squaring old_span_percent is because we want non-linear triggering of defragmenting old.  We don't really care about old fragmentation if old's span is small.  Doing too much defragmentation of OLD is shown to degrade performance when the defragmentation was really not "necessary".  With linear (not-squared) old_span_percent, we would be triggering almost twice as frequently at lower span values and less aggressively when old_span_percent exceeds its allowed span:
  // Trigger if old_span_percent is: 110% and density is below 82.5%
  //                                                      100%                                    75% (same at this trigger)
  //                                                       90%                                     67.5%
  //                                                       80%                                     60%
  //                                                       70%                                     56.25%
  //                                                       60%                                     45%
  //                                                       50%                                     37.5%                                      

As old span approaches the "target maximum span" it becomes increasingly important to defragment old.  On the other hand, if we totally ignore fragmentation until we have exceeded the limit, we have high probability of experiencing a STW Full GC due to humongous allocation failure.  This is because:

1. Old defragmentation generally requires many GC cycles, first to do OLD marking, and then to do the mixed evacuations.
2. Defragmenting mixed evacuations typically run slower than "traditional" mixed evacuations.  This is because a typical mixed evacuation only evacuates regions that have "abundant garbage".  But a defragmenting mixed evacuation may have to evacuate regions that are highly utilized.

One other reason to not totally ignore defragmentation until it has become "urgent" is because early investment in defragmenting can prevent the more severe forms of fragmentation from manifesting. When large numbers of regions with low utilization have been promoted in place (which s is one of the key causes of old fragmentation), it is easier to consolidate the live memory from multiple sparsely populated old-gen regions while they are still sparsely populated.  Otherwise, if we delay the defragmentation effort, it is likely that many of the sparsely populated regions become high utilized before we defragment, and then we'll have to copy a lot more data around in order to move regions out of the humongous region zone.

There is room for more extensive performance measurements and improved heuristics, I'm sure.

Reply:
Thank you for the detailed explanation. It makes more sense now.